docs/proposals/parameterized-config-management-plugins.md

```diff
@@ -0,0 +1,502 @@
+  * An Argo CD admin should be able to write an _advanced_ parameterized CMP server relying on thorough docs.
+    
+    Writing a custom CMP server might be preferable if the parameters announcement code gets too complex to be 
+    an inline shell script.
+* Easy to install
+  * Installing a simple CMP or even a CMP with a custom server should be intuitive and painless.
+* Easy to use
+  * Argo CD end-users (for example, developers) should be able to
+    1. View and set parameters in the Argo CD Application UI
+    2. See the parameters reflected in the Application manifest
+    3. Easily read/modify the generated parameters in the manifest (they should be structured in a way that's easy to read)
+  * CMPs should be able to announce parameters with more helpful interfaces than a simple text field.
+    * For example, image parameters should be presented using the same helpful interface as the one in Kustomize applications.
+* Future-proof
+  * Since the rich parameters UI is an important feature for config management tools, the parameter definition schema 
+    should be flexible enough to announce new _types_ of parameters so the UI can customize its presentation.
+* Backwards-compatible
+  * CMPs written before this enhancement should work fine after this enhancement is released.
+  * The UI should be able to handle unknown (new) parameter types. For example, if a plugin announces a parameter of 
+    type `date`, the UI should fall back to allowing text entry. The UI can then be enhanced to provide a better input
+    mechanism in a later release.
+* Proven with a rich demonstration
+  * The initial release of this feature should include a CMP implementation of the Helm config tool. This will
+    1. Serve as a rich example for others CMP developers to mimic
+    2. Allow us to decouple the Helm config management release cycle from the Argo release cycle
+    3. Allow us to work around [this bug](https://github.com/argoproj/argo-cd/issues/7291) without including the Helm 
+       SDK in the core Argo CD code
+  * The Helm CMP must be on-par with the native implementation.
+    1. It must present an equivalent parameters UI.
+    2. It must communicate errors back to the repo-server (and then the UI) the same as the native implementation.
+
+### Non-Goals
+
+We should not:
+* Re-implement config management tools as CMPs (besides Helm)
+
+## Proposal
+
+### Use cases
+
+#### Use case 1: building Argo CD without config management dependencies
+
+As an Argo CD developer, I would like to be able to build Argo CD without including the Helm SDK as a dependency.
+
+The Helm SDK includes the Kubernetes code base. That's a lot of code, and it will make builds unacceptably slow.
+
+#### Use case 2: writing CMPs with rich UI experiences
+
+As an Argo CD user, I would like to be able to parameterize manifests built by a CMP.
+
+For example, if the Argo CD administrator has installed a CMP which applies a last-mile kustomize overlay to a Helm
```

Comment:
> which applies a last-mile kustomize overlay to a Helm repo

Not sure I understand this use-case. Does it mean that we are mixing Helm Charts and Kustomize in the same application?

Reply:
@leoluz potentially. The specific use case I'm thinking of is when someone wants to use an external Helm chart (say, minio), but then they want to add a secret or apply an annotation or some other minor change.

This use case could be solved by creating an internal repo (https://github.company.com/org/kustomized-minio.git) which pulls in the minio Helm chart (using the kustomize `helmCharts` feature) but then applies a kustomize overlay.

You can do that with Argo CD today. But then you won't be able to set any of the minio Helm parameters in the Argo CD UI.

A CMP with robust parameter discovery could 1) parse the kustomization.yaml to find all the `helmCharts`, 2) perform parameter discovery on each Helm chart, and 3) announce each Helm chart's parameters in their own "section" in the parameter announcement.